modules/node/src/appRegistry/appRegistry.service.ts

```diff
@@ -41,25 +43,103 @@ export class AppRegistryService {
     private readonly swapRateService: SwapRateService,
     private readonly channelService: ChannelService,
     private readonly transferService: TransferService,
+    private readonly configService: ConfigService,
     private readonly appRegistryRepository: AppRegistryRepository,
     private readonly channelRepository: ChannelRepository,
-    private readonly configService: ConfigService,
     private readonly linkedTransferRepository: LinkedTransferRepository,
+    @Inject(MessagingClientProviderId) private readonly messagingClient: ClientProxy,
   ) {}
 
-  /**
-   * Validate apps that are proposed to be installed on the node by clients and
-   * accept or reject the install.
-   * @param data Data from CF event PROPOSE_INSTALL
-   */
-  async allowOrReject(data: ProposeMessage): Promise<AppRegistry | void> {
-    const registryAppInfo = await this.verifyAppProposal(data.data, data.from);
-    if (registryAppInfo.name === CoinBalanceRefundApp) {
-      logger.log(`Not installing coin balance refund app, returning registry information`);
-      return registryAppInfo;
-    }
-    await this.cfCoreService.installApp(data.data.appInstanceId);
-    return registryAppInfo;
+  async validateAndInstallOrReject(
+    appInstanceId: string,
+    proposeInstallParams: CFCoreTypes.ProposeInstallParams,
+    from: string,
+  ): Promise<void> {
+    let registryAppInfo: AppRegistry;
+    try {
+      registryAppInfo = await this.verifyAppProposal(proposeInstallParams, from);
+      if (registryAppInfo.name !== CoinBalanceRefundApp) {
+        await this.cfCoreService.installApp(appInstanceId);
+      } else {
+        logger.log(`Not installing coin balance refund app, returning registry information`);
+      }
+    } catch (e) {
+      if (!e.message.includes(`Node has insufficient balance`)) {
+        logger.warn(`App install failed, . Error: ${e.stack || e.message}`);
+        await this.cfCoreService.rejectInstallApp(appInstanceId);
+        return;
+      }
+      // try to collateralize, and re-install
+      const tx = await this.channelService.rebalance(
+        from,
+        proposeInstallParams.responderDepositTokenAddress,
+        RebalanceType.COLLATERALIZE,
+        bigNumberify(proposeInstallParams.responderDeposit),
+      );
+      if (tx) {
+        await tx.wait();
+      }
+      try {
+        await this.cfCoreService.installApp(appInstanceId);
+      } catch (e) {
+        logger.warn(`App install failed, . Error: ${e.stack || e.message}`);
+        await this.cfCoreService.rejectInstallApp(appInstanceId);
+      }
```

Comment:
won't this still call `this.runPostInstallTasks` after the app is rejected?

Reply:
Yikes, need a return there. Good catch.